torch/distributed/_shard/sharding_spec/api.py

```diff
@@ -195,31 +193,69 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
             memory_format=torch.contiguous_format,
             pin_memory=tensor.is_pinned()
         )
+        current_rank = dist.get_rank(process_group)
         tensor_meta = self.build_metadata(tensor.size(), tensor_properties)
         local_shards = []
+        local_tensor = None
+        local_metadata = None
+        tensors_to_scatter = [None] * dist.get_world_size(process_group)
 
-        current_rank = dist.get_rank(process_group)
-        # Scatter the shards (use broadcast since NCCL doesn't support scatter, this is very inefficient).
-        dist.broadcast(tensor, src=src_rank, group=process_group)
+        sharding_dim_size = tensor.size()[self.dim]  # type: ignore[index]
+        chunks = len(self.placements)
+        split_size = get_split_size(sharding_dim_size, chunks)
+        scatter_shape = list(tensor.size())
+        scatter_shape[self.dim] = split_size  # type: ignore[index]
 
         for shard_meta in tensor_meta.shards_metadata:
             rank, device = _parse_and_validate_remote_device(process_group, shard_meta.placement)
-            if rank == current_rank:
+            if current_rank == src_rank:
                 # Reshape to get shard for this rank and we don't want autograd
                 # recording here for the narrow op and 'local_shard' should be a
                 # leaf variable in the autograd graph.
-                local_tensor = narrow_tensor(tensor, shard_meta).clone().detach().contiguous()
-
-                # Sync requires_grad to local_shard.
-                local_tensor.requires_grad = tensor.requires_grad
-                local_shards.append(
-                    Shard(
-                        tensor=local_tensor,
-                        metadata=shard_meta
-                    )
-                )
+                narrowed_tensor = narrow_tensor(tensor, shard_meta)
+                if shard_meta.shard_sizes[self.dim] < split_size:  # type: ignore[index]
+                    # for the last shard that might be smaller to other shards
+                    # resize the narrowed tensor to the same size and use it for
+                    # the scatter collective as dist.scatter requires same size
+                    # inputs on every rank
+                    tensor_to_scatter = narrowed_tensor.detach().clone().resize_(scatter_shape)
+                else:
+                    tensor_to_scatter = narrowed_tensor.detach().clone().contiguous()
+
+                tensors_to_scatter[rank] = tensor_to_scatter
+
+            if current_rank == rank:
```

Comment:
nit: This is the same condition as above and can be merged?

Reply:
Hmm so it's a bit different, the first condition is `current_rank=src_rank` where it collects all the shards to scatter (only on src_rank), the second condition is `current_rank = rank` where it check if we are in the rank of the current `shard_metadata`.